raft.addPeer succeeds, but new node panics when trying to add new raft peer with same raftport

Its possible to mis-configure Quorum running in raft mode in such a way that if you manually add a peer with the same IP:PORT:RAFTPORT configuration as an existing peer, new peer fails with the following panic error

```
raft2018/06/13 22:11:14 tocommit(14) is out of range [lastIndex(0)]. Was the raft log corrupted, truncated, or lost?
panic: tocommit(14) is out of range [lastIndex(0)]. Was the raft log corrupted, truncated, or lost?

goroutine 56 [running]:
log.(*Logger).Panicf(0xc4201de140, 0x10ab169, 0x5d, 0xc4215e86e0, 0x2, 0x2)
        /usr/local/go/src/log/log.go:219 +0xdb
github.com/ethereum/go-ethereum/vendor/github.com/coreos/etcd/raft.(*DefaultLogger).Panicf(0x1abbf50, 0x10ab169, 0x5d, 0xc4215e86e0, 0x2, 0x2)
        /home/vagrant/quorum/build/_workspace/src/github.com/ethereum/go-ethereum/vendor/github.com/coreos/etcd/raft/logger.go:121 +0x60
github.com/ethereum/go-ethereum/vendor/github.com/coreos/etcd/raft.(*raftLog).commitTo(0xc4203dc850, 0xe)
        /home/vagrant/quorum/build/_workspace/src/github.com/ethereum/go-ethereum/vendor/github.com/coreos/etcd/raft/log.go:191 +0x15c
github.com/ethereum/go-ethereum/vendor/github.com/coreos/etcd/raft.(*raft).handleHeartbeat(0xc42033be00, 0x8, 0xa, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/vagrant/quorum/build/_workspace/src/github.com/ethereum/go-ethereum/vendor/github.com/coreos/etcd/raft/raft.go:1095 +0x54
github.com/ethereum/go-ethereum/vendor/github.com/coreos/etcd/raft.stepFollower(0xc42033be00, 0x8, 0xa, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/vagrant/quorum/build/_workspace/src/github.com/ethereum/go-ethereum/vendor/github.com/coreos/etcd/raft/raft.go:1041 +0x2a7
github.com/ethereum/go-ethereum/vendor/github.com/coreos/etcd/raft.(*raft).Step(0xc42033be00, 0x8, 0xa, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/vagrant/quorum/build/_workspace/src/github.com/ethereum/go-ethereum/vendor/github.com/coreos/etcd/raft/raft.go:778 +0x1051
github.com/ethereum/go-ethereum/vendor/github.com/coreos/etcd/raft.(*node).run(0xc4203dc930, 0xc42033be00)
        /home/vagrant/quorum/build/_workspace/src/github.com/ethereum/go-ethereum/vendor/github.com/coreos/etcd/raft/node.go:346 +0x1189
created by github.com/ethereum/go-ethereum/vendor/github.com/coreos/etcd/raft.StartNode
        /home/vagrant/quorum/build/_workspace/src/github.com/ethereum/go-ethereum/vendor/github.com/coreos/etcd/raft/node.go:215 +0x3db
```

Tested using quorum-examples. To repeat, initialize vagrant quorum-examples, then raft-init, then raft-start. Now, configure a new node like so:
```
geth --datadir qdata/dd-1000 init genesis.json
```

Next, start it up to get its enode info:
```
geth --datadir qdata/dd-1000 --nodiscover --raft --rpc --rpcaddr 0.0.0.0 --rpcapi admin,db,eth,debug,miner,net,shh,txpool,personal,web3,quorum --emitcheckpoints --raftport 55555 --rpcport 22222 --port 55550 console
```

This will print something like this:
```
Fatal: Raft-based consensus requires either (1) an initial peers list (in static-nodes.json) including this enode hash (97708573cdac7c4d4a8dc14e4e7a5b6811d68bcce4f573ac562a41a8ade6b1448d59a5008d4c4081a789bd77cf1681395cb62910404f7bc9ab8bddfe5e034a0f), or (2) the flag --raftjoinexisting RAFT_ID, where RAFT_ID has been issued by an existing cluster member calling `raft.addPeer(ENODE_ID)` with an enode ID containing this node's enode hash.
```

Now, lets add this to our dd1 via raft.addPeer. This will print raft id that we should use, in my case *8*. Restarting my node with joining raft consensus works fine:
```
geth --datadir qdata/dd-1000 --nodiscover --raft --rpc --rpcaddr 0.0.0.0 --rpcapi admin,db,eth,debug,miner,net,shh,txpool,personal,web3,quorum --emitcheckpoints --raftport 55555 --rpcport 22222 --port 55550 --raftjoinexisting 8 console
```

Now, for the error. Stop the new node, and repeat the steps above -- the only things that should change would be enode address and raftjoinexisting should be bumped up to *9*. Everything will work fine until the last step, when new node is supposed to connect and sync up.

Note that if raft.removePeer(OLD_PEER_NUMBER) is executed on dd1, then the second run works correctly. We should either clean up raft peers correctly or otherwise warn client that new node is mis-configured.

--- a/raft/handler.go
+++ b/raft/handler.go
@@ -274,17 +274,28 @@ func (pm *ProtocolManager) isRaftIdUsed(raftId uint16) bool {
 	return pm.peers[raftId] != nil
 }
 
-func (pm *ProtocolManager) isP2pNodeInCluster(node *discover.Node) bool {
+func (pm *ProtocolManager) isNodeAlreadyInCluster(node *discover.Node) error {
 	pm.mu.RLock()
 	defer pm.mu.RUnlock()
 
 	for _, peer := range pm.peers {
-		if peer.p2pNode.ID == node.ID {
-			return true
+		peerRaftId := peer.address.RaftId
+		peerNode := peer.p2pNode
+
+		if peerNode.ID == node.ID {
+			return fmt.Errorf("node with this enode has already been added to the cluster: %v", node.ID)
+		}
+
+		if peerNode.IP.Equal(node.IP) {
+			if peerNode.TCP == node.TCP {
+				return fmt.Errorf("existing node %v with raft ID %v is already using eth p2p at %v:%v", peerNode.ID, peerRaftId, node.IP, node.TCP)
+			} else if peer.address.RaftPort == node.RaftPort {
+				return fmt.Errorf("existing node %v with raft ID %v is already using raft at %v:%v", peerNode.ID, peerRaftId, node.IP, node.RaftPort)
+			}
 		}
 	}
 
-	return false
+	return nil
 }
 
 func (pm *ProtocolManager) ProposeNewPeer(enodeId string) (uint16, error) {
@@ -293,16 +304,16 @@ func (pm *ProtocolManager) ProposeNewPeer(enodeId string) (uint16, error) {
 		return 0, err
 	}
 
-	if pm.isP2pNodeInCluster(node) {
-		return 0, fmt.Errorf("node is already in the cluster: %v", enodeId)
-	}
-
 	if len(node.IP) != 4 {
 		return 0, fmt.Errorf("expected IPv4 address (with length 4), but got IP of length %v", len(node.IP))
 	}
 
 	if !node.HasRaftPort() {
 		return 0, fmt.Errorf("enodeId is missing raftport querystring parameter: %v", enodeId)
+	}
+
+	if err := pm.isNodeAlreadyInCluster(node); err != nil {
+		return 0, err
 	}
 
 	raftId := pm.nextRaftId()

